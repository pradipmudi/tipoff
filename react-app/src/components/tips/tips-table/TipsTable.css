#tips-table thead {
    background-color: #e1f4f3;

}

#tips-table tbody{
    background-color: #ffffff;
}

.view-tip-icon {
    text-align: center !important;
}

.error{
    background-color: rgb(230, 188, 152) !important;
}

.warning{
    background-color: #f3f3b6  !important;
}

.success {
    background-color:rgb(204, 233, 178)  !important;
}

.unhandled {
    background-color:rgb(202, 136, 136)  !important;
}

.info {
    background-color:#a7c6d3  !important;
}

