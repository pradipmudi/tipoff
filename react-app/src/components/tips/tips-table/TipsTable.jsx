import React from 'react';
import { useHistory } from 'react-router-dom';
import Moment from 'react-moment';
import { Table, Content, Icon } from 'bloomer';

import UserAgentMini from '../../useragent/UserAgentMini';
import Tags from '../../common/tags/Tags';

import './TipsTable.css';

const TipsTable = ({tips}) => {
    
    const history = useHistory();

    const viewTip = (e, tipId) => { 
        e.preventDefault();
        history.push({
            pathname: '/tip/'+tipId
        });
    }
    return (
        <Content>
            <Table isBordered isStriped  id="tips-table">
                <thead >
                    <tr className = "header-row" >
                        <th className="is-hidden-touch">Event Type</th>
                        <th>UA</th>
                        <th>Event Title</th>
                        <th className="is-hidden-touch">Tags</th>
                        <th>Created On</th>
                        <th className="is-hidden-touch">View</th>
                    </tr>
                </thead>
                {tips && tips.length ? (
                    <tbody >
                    {tips.map((tip) => 
                        <tr key={tip.tip_id} 
                                className = {'tip-row '+tip.event_type} 
                                onClick={(e)=>{viewTip(e, tip.tip_id)}}
                                style={{cursor:"pointer"}}>
                            <td className="is-hidden-touch">{tip.event_type}</td>
                            <td><UserAgentMini ua={tip.user_agent} ></UserAgentMini></td>
                            <td>{tip.event_title}</td>
                            <td className="is-hidden-touch"><Tags tags={tip.tags} ></Tags> </td>
                            <td><Moment fromNow>{tip.created_on}</Moment></td>
                            <td className="is-hidden-touch view-tip-icon" ><Icon className='fas fa-file-alt' /></td>
                        </tr>
                    )}
                    </tbody>
                ) : (
                    <tbody>
                        <tr>
                            <td colSpan="6">This project hasn't received any Tips yet. You can refer to the Setup Guide to get started.</td>
                        </tr>
                    </tbody>
                )}
            </Table>
        </Content>
    )
};

export default TipsTable;
