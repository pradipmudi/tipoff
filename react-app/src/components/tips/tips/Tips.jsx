import React, { useState, useEffect } from 'react';
import { useSelector, useDispatch, shallowEqual } from 'react-redux';
import { useRouteMatch, useLocation } from 'react-router-dom';
import { Content  } from 'bloomer';
import axios from 'axios';

import { setTips } from '../../../redux/tips/tips-actions';
import { 
	selectTips, 
	selectTipsItems, 
	selectTipsPageNo, 
	selectTipsSearchTerm 
} from '../../../redux/tips/tips-selectors';

import TipsTable from '../tips-table/TipsTable';
import PageTitle from '../../common/page-title/PageTitle';
import TipsSerach from '../tips-search/TipsSearch';
import TipsLoadMore from '../tips-load-more/TipsLoadMore';


import './Tips.css';

export const Tips = () =>  {
	//declare all the hooks
	const match = useRouteMatch();
	const dispatch = useDispatch();
	const tips = useSelector(selectTips, shallowEqual);
	var location = useLocation();
	
	const [loading, setLoading] = useState(false);
	const [showLoadMore, setShowLoadMore] = useState(true);
	const page = useSelector(selectTipsPageNo, shallowEqual);
	const items = useSelector(selectTipsItems, shallowEqual);
	const searchTerm = useSelector(selectTipsSearchTerm, shallowEqual);
	const id = match.params.id;
	const pageTitle = location.state ? location.state.projectName : id;
	//TODO get project name from db

	

	useEffect(() => {
		const minTips = (page-1) * items + 1;
		const maxTips = (page) * items;
		let currentTips = tips.length;

		const fetchNewTips = () => {
			
			if (minTips <= currentTips && currentTips <=maxTips ){
				return false;
			} else {
				return true;
			}
		}

		const isShowMoreViewable = () => {
			let viewable = currentTips === maxTips ? true : false;
			return viewable;
		}

		const getTips = async () => {
			
			if (fetchNewTips()){
				setLoading(true);
				axios.get(`/api/v1/tips?pageNo=${page}&items=${items}&projectId=${id}&searchTerm=${searchTerm}`)
					.then((resp) => {
						if(resp && resp.data && resp.data.success){
							const respTips = resp.data.data;
							let newTips = tips.concat(respTips);
							dispatch(setTips(newTips));
							currentTips = currentTips + respTips.length;
						}
					})
					.catch(err => { console.log(err) })
					.finally(()=> { 
						setLoading(false);
						setShowLoadMore(isShowMoreViewable())
					});
			} else {
				setShowLoadMore(isShowMoreViewable());
			}
			
		}
		getTips();
	}, [page, items, dispatch, id, searchTerm, tips]);

	

	return (
		<Content>
			<PageTitle pageTitle={pageTitle} loading={loading} />
			<TipsSerach ></TipsSerach>
			<TipsTable tips={tips} />
			{ showLoadMore && <TipsLoadMore /> }
		</Content>			
	);
}

export default Tips;
