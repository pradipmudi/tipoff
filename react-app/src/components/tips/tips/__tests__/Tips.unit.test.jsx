import React from 'react';
import {shallow} from 'enzyme';
import axios from 'axios';
import renderer from 'react-test-renderer';


import {Tips}  from '../Tips';
import TipsTable from '../../tips-table/TipsTable';

const mockTips = {
    data: {  
        success: true,
        data: [{
        
            "project_id" : "IotWorx",
            "event_type" : "info",
            "event_title" : "Fetching news with page 2",
            "created_on" : "2019-12-29T18:07:40.409Z",
            "note" : "Bedfordshire Small",
            "tags" : [
                    "jobs",
                    "news",
                    "resources"
            ],
            "user" : {
                    "user_id" : "Yessenia.Jenkins@gmail.com"
            },
            "user_agent" : {
                    "ua" : "Mozilla/5.0 (Windows NT 10.0; WOW64; rv:68.0) Gecko/20100101 Firefox/68.0",
                    "browser" : {
                            "name" : "Firefox",
                            "version" : "68.0",
                            "major" : "68"
                    },
                    "engine" : {
                            "name" : "Gecko",
                            "version" : "68.0"
                    },
                    "os" : {
                            "name" : "Windows",
                            "version" : "10"
                    },
                    "cpu" : {
                            "architecture" : "amd64"
                    }
            },
            "tip_id" : "ck5bgpv300000i95u4gxogtp9",
            "__v" : 0
        }]
    }
};


describe("Tips Component", () => {

    const tipsState = {
        tips: mockTips.data.data,
        page: 1,
        items: 100
    }

    const setTips = () => {}

    beforeEach(() => {
        axios.get.mockImplementationOnce(() =>
            Promise.resolve(mockTips)
        );
    });

    afterEach(() => {
        jest.restoreAllMocks();
    });

    it("matches snapshot", () => {
        const tree = renderer.create(<Tips tipsState={tipsState} setTips={setTips} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('calls `getTips` function', async (done) => {
        const tips = shallow (<Tips tipsState={tipsState} setTips={setTips}/>);
        tips.instance().componentDidMount().then(() => {
            expect(axios.get).toHaveBeenCalled();
            expect(tips.find('PageTitle').length).toEqual(1);
            expect(tips.find('TipsTable').length).toEqual(1);
            expect(tips.find('TipsTable').props().tips).toEqual(mockTips.data.data);
            done();
        });
    });

});