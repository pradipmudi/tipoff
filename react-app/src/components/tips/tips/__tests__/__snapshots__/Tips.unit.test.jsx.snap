// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tips Component matches snapshot 1`] = `
<div
  className="content"
>
  <h1
    className="title is-2 pageTitle"
  >
    Project1
  </h1>
  <div
    className="content"
  >
    <table
      className="table is-bordered is-narrow is-striped"
      id="tips-table"
    >
      <thead>
        <tr
          className="header-row"
        >
          <th>
            Event Type
          </th>
          <th>
            Created On
          </th>
          <th>
            Event Title
          </th>
          <th>
            Tags
          </th>
          <th>
            User Agent
          </th>
          <th>
            User ID
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          className="tip-row info"
        >
          <td>
            info
          </td>
          <td>
            <time
              dateTime={"2019-12-29T18:07:40.409Z"}
            >
              a month ago
            </time>
          </td>
          <td>
            Fetching news with page 2
          </td>
          <td>
            <div>
              <span>
                <span
                  className="tag dark-tag"
                  style={
                    Object {
                      "margin": "5px",
                    }
                  }
                >
                  jobs
                </span>
                <span
                  className="tag dark-tag"
                  style={
                    Object {
                      "margin": "5px",
                    }
                  }
                >
                  news
                </span>
                <span
                  className="tag dark-tag"
                  style={
                    Object {
                      "margin": "5px",
                    }
                  }
                >
                  resources
                </span>
              </span>
            </div>
             
          </td>
          <td>
            <span>
              <span>
                <span
                  className="icon"
                  title="Firefox:68"
                >
                  <i
                    aria-hidden="true"
                    className="fab fa-firefox"
                  />
                </span>
              </span>
              <span>
                <span
                  className="icon"
                  title="Windows:10"
                >
                  <i
                    aria-hidden="true"
                    className="fab fa-windows"
                  />
                </span>
              </span>
              <span>
                <span
                  className="icon"
                >
                  <i
                    aria-hidden="true"
                    className="fas fa-laptop"
                  />
                </span>
              </span>
            </span>
          </td>
          <td>
            Yessenia.Jenkins@gmail.com
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
